plenum/server/replica.py

```diff
@@ -996,6 +1002,7 @@ def processPrepare(self, prepare: Prepare, sender: str) -> None:
             return
         try:
             if self.validatePrepare(prepare, sender):
+                self._setup_last_ordered_for_non_master(is_catchup=False)
```

Comment:
If validation of `Prepare` succeeds then this means that the 3PC-batch is next one (i.e. with the next `ppSeqNo`) and in such the case we do not need to adjust `last_ordered_3pc`.

This call must be moved to `validate_prepare` method to the point(s) after the call(s) of `enqueue_prepare` (since the prepared certificate may become completed only after addition of an incoming message to `prePreparesPendingPrevPP` or `preparesWaitingForPrePrepare`).

In case the prepared certificate is completed and last_ordered_3pc is adjusted for it then actual processing of the pending messages (from `prePreparesPendingPrevPP` and `preparesWaitingForPrePrepare`) will be initiated by `dequeue_pre_prepares` call when `Replica.serviceQueues` procedure is executed next time.

Reply:
Yes, it's a problem. Should we move _setup_last_ordered_for_non_master before enqueue_prepare() for non master replica in validation?